flang/lib/Lower/PFTBuilder.cpp

```diff
@@ -317,10 +317,12 @@ class PFTBuilder {
       // construct region must have an exit target inside the region.
       Fortran::lower::pft::EvaluationList &evaluationList =
           *eval->evaluationList;
-      if (!evaluationList.empty() && evaluationList.back().isConstruct()) {
+      if (!evaluationList.empty() && (evaluationList.back().isConstruct() ||
+                                      evaluationList.back().isDirective())) {
         static const parser::ContinueStmt exitTarget{};
         addEvaluation(
-            lower::pft::Evaluation{exitTarget, pftParentStack.back(), {}, {}});
+            lower::pft::Evaluation{exitTarget, pftParentStack.back(), {}, {}})
+            .isNewBlock = true;
```

Comment:
Branch analysis should automatically set the `isNewBlock` flag, at least for looping constructs.  Are there "straight-line" OpenMP constructs that aren't automatically setting this flag?  (If yes, there is another option for getting an "exit" for it.)  For looping constructs, if the `ContinueStmt` is not being marked as a new block, it is probably in the wrong place, perhaps with respect to an `OmpEndLoopDirective` eval.

Reply:
For looping constructs, the `ContinueStmt` is marked as a new block. It is for the outer parallel construct that the `ContinueStmt` is not marked as a new block. Due to this, previously, they end up sharing the same block when `createEmptyRegionBlocks` is called for both.

Reply:
If I understand how code should look for parallel regions, not all inserted `ContinueStmt`s are branch targets.  Unstructured constructs need not have branches that exit the construct.  A loop could be unstructured due to the presence of a simple `goto` entirely local to the loop body, or some nested construct with an `ExitStmt`.  I believe all basic block recognition is currently local to PFT routine `analyzeBranches`.

That said, if these `ContinueStmt`s must always be in their own (unstructured) basic block, or you can at least determine that forcing such a block all the time is not a correctness problem, then it might be ok to initially force creation of such blocks all the time, with the option of going back later to sharpen up the code to only set it when needed.  The `-no-structured-fir` flag should make that testing a little more robust.

It should be easier to investigate the problem once PR #1399 is available; thanks for splitting that out.

Reply:
I have now removed the requirement for `ContinueStmt`s to be always in their own basic block. This became possible after restricting `createEmptyRegionBlocks` to not recurse into nested OpenMPConstructs/Directives.